File: app/src/main/java/com/sasfmlzr/findfilm/fragment/currentfilm/CurrentFilmViewModel.java
package com.sasfmlzr.findfilm.fragment.currentfilm;

import android.databinding.BaseObservable;
import android.databinding.Bindable;
import android.databinding.ObservableBoolean;
import android.databinding.ObservableField;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;

import com.android.databinding.library.baseAdapters.BR;
import com.sasfmlzr.findfilm.model.RetrofitSingleton;
import com.sasfmlzr.findfilm.request.CurrentMovieRequest;
import com.sasfmlzr.findfilm.request.FindFilmApi;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.sasfmlzr.findfilm.model.SystemSettings.API_KEY;
import static com.sasfmlzr.findfilm.model.SystemSettings.LANGUAGE;
import static com.sasfmlzr.findfilm.model.SystemSettings.URL_IMAGE_500PX;

public class CurrentFilmViewModel extends BaseObservable {
    private static final ObservableField<String> materialMessage = new ObservableField<>("Buy tickets pressed");
    private static final ObservableField<String> videoMessage = new ObservableField<>("Play video pressed");

    public final ObservableBoolean dataLoading = new ObservableBoolean(false);
    private int idFilm;

    @Bindable
    private CurrentMovieRequest currentMovieField;

    @Bindable
    private String toastMessage = null;

    public interface FilmLoaded {
        void isLoaded(CurrentMovieRequest currentMovieRequest);
    }

    public interface DownloadImage {
        void isDownloaded(CurrentMovieRequest currentMovieRequest);
    }

    public void start(int filmId) {
        this.idFilm = filmId;
        loadFilm();
    }

    public CurrentMovieRequest getCurrentMovieField() {
        return currentMovieField;
    }

    private void setCurrentMovieField(CurrentMovieRequest currentMovieField) {
        this.currentMovieField = currentMovieField;
        notifyPropertyChanged(BR.currentMovieField);
    }

    private void loadFilm() {
        DownloadImage imageDownloadCallback = (film) -> {
            currentMovieField.setBackdropBitmap(film.getBackdropBitmap());
            dataLoading.set(true);
        };

        FilmLoaded filmLoadCallback = (currentMovieRequest) -> {
            setCurrentMovieField(currentMovieRequest);

            String url;
            if (currentMovieRequest.getBackdropPath() == null) {
                url = URL_IMAGE_500PX + currentMovieRequest.getPosterPath();
            } else {
                url = URL_IMAGE_500PX + currentMovieRequest.getBackdropPath();
            }
            Target target = new Target() {
                @Override
                public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                    currentMovieRequest.setBackdropBitmap(bitmap);
                    imageDownloadCallback.isDownloaded(currentMovieRequest);
                }

                @Override
                public void onBitmapFailed(Exception e, Drawable errorDrawable) {

                }

                @Override
                public void onPrepareLoad(Drawable placeHolderDrawable) {

                }
            };
            Picasso.get().load(url).into(target);
        };

        FindFilmApi findFilmApi = RetrofitSingleton.getFindFilmApi();
        findFilmApi.getCurrentMovie(idFilm, API_KEY, LANGUAGE)
                .enqueue(new Callback<CurrentMovieRequest>() {
                    @Override
                    public void onResponse(Call<CurrentMovieRequest> call, Response<CurrentMovieRequest> response) {
                        filmLoadCallback.isLoaded(response.body());
                    }

                    @Override
                    public void onFailure(Call<CurrentMovieRequest> call, Throwable t) {

                    }
                });
    }
    
    /**
     * Toast message methods
     */
    public String getToastMessage() {
        return toastMessage;
    }

    private void setToastMessage(ObservableField<String> toastMessage) {
        this.toastMessage = toastMessage.get();
        notifyPropertyChanged(BR.toastMessage);
    }

    public void materialButtonClicked() {
        setToastMessage(materialMessage);
    }

    public void videoButtonClicked() {
        setToastMessage(videoMessage);
    }
}

File: app/src/main/java/com/sasfmlzr/findfilm/fragment/currentfilm/CurrentFilmViewModel.kt
package com.sasfmlzr.findfilm.fragment.currentfilm

import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import androidx.databinding.BaseObservable
import androidx.databinding.Bindable
import androidx.databinding.ObservableBoolean
import androidx.databinding.library.baseAdapters.BR
import androidx.lifecycle.MutableLiveData
import com.sasfmlzr.findfilm.model.RetrofitSingleton
import com.sasfmlzr.findfilm.model.SystemSettings
import com.sasfmlzr.findfilm.request.CurrentMovieRequest
import com.sasfmlzr.findfilm.request.FindFilmApi
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class CurrentFilmViewModel : BaseObservable() {
    private var idFilm = 0

    val dataLoading: ObservableBoolean = ObservableBoolean(false)

    val currentMovieField = MutableLiveData<CurrentMovieRequest>()

    val poster = MutableLiveData<Bitmap>()

    /**
     * Toast message methods
     */
    @Bindable
    var toastMessage: String? = null
        private set(toastMessage) {
            field = toastMessage!!.toString()
            notifyPropertyChanged(BR.toastMessage)
        }

    interface FilmLoaded {
        fun isLoaded(currentMovieRequest: CurrentMovieRequest)
    }

    interface DownloadImage {
        fun isDownloaded(currentMovieRequest: CurrentMovieRequest)
    }

    fun start(filmId: Int) {
        idFilm = filmId
        loadFilm()
    }

    private fun loadFilm() {
        val imageDownloadCallback = object : DownloadImage {
            override fun isDownloaded(currentMovieRequest: CurrentMovieRequest) {
                poster.value =currentMovieRequest.backdropBitmap
                dataLoading.set(true)
            }
        }
        val filmLoadCallback = object : FilmLoaded {
            override fun isLoaded(currentMovieRequest: CurrentMovieRequest) {
                currentMovieField.value = currentMovieRequest
                val url: String = if (currentMovieRequest.backdropPath == null) {
                    SystemSettings.URL_IMAGE_500PX + currentMovieRequest.posterPath
                } else {
                    SystemSettings.URL_IMAGE_500PX + currentMovieRequest.backdropPath
                }
                val target: Target = object : Target {
                    override fun onBitmapLoaded(bitmap: Bitmap?, from: Picasso.LoadedFrom?) {
                        currentMovieRequest.backdropBitmap =bitmap
                        imageDownloadCallback.isDownloaded(currentMovieRequest)
                    }

                    override fun onBitmapFailed(e: Exception?, errorDrawable: Drawable?) {}
                    override fun onPrepareLoad(placeHolderDrawable: Drawable?) {

                    }
                }
                Picasso.get().load(url).into(target)
            }
        }
        val findFilmApi: FindFilmApi = RetrofitSingleton.findFilmApi!!
        findFilmApi.getCurrentMovie(idFilm, SystemSettings.API_KEY, SystemSettings.LANGUAGE).enqueue(object : Callback<CurrentMovieRequest> {
                override fun onResponse(
                    call: Call<CurrentMovieRequest>,
                    response: Response<CurrentMovieRequest>
                ) {
                    filmLoadCallback.isLoaded(response.body()!!)
                }

                override fun onFailure(call: Call<CurrentMovieRequest?>, t: Throwable) {}
            })
    }

    fun materialButtonClicked() {
        toastMessage = "Buy tickets pressed"
    }

    fun videoButtonClicked() {
        toastMessage = "Play video pressed"
    }
}
